src/modules/auth/login.controller.ts

```diff
@@ -76,6 +83,47 @@ export class LoginController {
     res.send({ token, refreshToken });
   }
 
+  @Post('login/siwe/initiate')
+  async initiateSiweLogin(@Res() res: Response) {
+    const nonce = generateNonce();
+    const expire = this.configService.get<string>('SIWE_NONCE_EXPIRES_IN');
+    const expireInSec = parseDuration(expire) / 1000;
+    await this.redis.set(nonce, 'true', { EX: expireInSec });
+    res.send({ nonce });
+  }
+
+  @UseGuards(LoginGuard)
+  @ApiBody({ type: SiweReqPayload })
+  @Post('login/siwe/verify')
+  async loginSiwe(
+    @Req() req: Request,
+    @Res() res: Response,
+    @Body({
+      transform: ({ message }) => ({
+        message: new SiweMessage(message),
+      }),
+    })
+    { message }: SiweReqPayload
+  ) {
+    const { nonce } = message;
+
+    if (!(await this.redis.exists(nonce))) {
+      throw new UnauthorizedException(
+        'Authentication with SIWE was not initiated'
+      );
+    }
+
+    if ((await this.redis.get(nonce)) === 'false') {
+      throw new UnauthorizedException(
+        'Authentication with SIWE has completed already'
+      );
+    }
+
+    await this.redis.set(nonce, 'false');
```

Comment:
With this implementation, i think it would technically be possible that a nonce is reused. The sequence would be
1. a client submits a `verify` request to serverA and serverB
2. serverA gets the value and sees that it is not false
3. serverB gets the value and sees that it is not false
4. serverA sets the value to false
5. server B sets the value to false
6. both servers return the login response
I think that this isn't such a big problem but it could potentially lead to unexpected behaviour or security risk.
Perhaps a way to improve this would be to use a SET with the GET option (https://redis.io/commands/set/). Something like:
```suggestion
    if ((await this.redis.get(nonce)) === 'false') {
      throw new UnauthorizedException(
        'Authentication with SIWE has completed already'
      );
    }

    await this.redis.set(nonce, 'false');
    // if previousVal was true, then it would mean it was the first
    if (previousVal === 'true') {
       return await this.login(req, res);
    }
    else if (previousVal === 'false') {
      throw new UnauthorizedException(
        'Authentication with SIWE has completed already'
      );
   }
   else {
     // throw an error
   }
```

Reply:
You right. Thanks for noticing this case

Reply:
the above means that if `this.login(req, res)` fails for some reason, another nonce needs to be generated. Correct?